tsl/src/continuous_aggs/invalidation.c

```diff
@@ -130,13 +133,62 @@ create_invalidation_tup(const TupleDesc tupdesc, int32 cagg_hyper_id, int64 modt
 	return heap_form_tuple(tupdesc, values, isnull);
 }
 
+/*
+ * Add an entry to the continuous aggregate invalidation log.
+ */
+void
+invalidation_cagg_log_add_entry(int32 cagg_hyper_id, int64 modtime, int64 start, int64 end)
+{
+	Relation rel = open_invalidation_log(LOG_CAGG, RowExclusiveLock);
+	CatalogSecurityContext sec_ctx;
+	HeapTuple tuple;
+
+	tuple = create_invalidation_tup(RelationGetDescr(rel), cagg_hyper_id, modtime, start, end);
+	ts_catalog_database_info_become_owner(ts_catalog_database_info_get(), &sec_ctx);
+	ts_catalog_insert_only(rel, tuple);
+	ts_catalog_restore_user(&sec_ctx);
+	heap_freetuple(tuple);
+	table_close(rel, NoLock);
+}
+
 typedef enum InvalidationResult
 {
 	INVAL_NOMATCH,
 	INVAL_DELETE,
 	INVAL_CUT,
 } InvalidationResult;
 
+static void
+save_invalidation(const CaggInvalidationState *state, int64 modification_time,
+				  int64 lowest_modified_value, int64 greatest_modified_value)
+{
+	int32 cagg_hyper_id = state->cagg.data.mat_hypertable_id;
+	TupleDesc tupdesc = RelationGetDescr(state->cagg_log_rel);
+	HeapTuple refresh_tup;
+
+	refresh_tup = create_invalidation_tup(tupdesc,
+										  cagg_hyper_id,
+										  modification_time,
+										  lowest_modified_value,
+										  greatest_modified_value);
+	tuplestore_puttuple(state->invalidations, refresh_tup);
+	heap_freetuple(refresh_tup);
+}
```

Comment:
For all uses of this function, you have an `Invalidation` instance to pass in. I suggest that you do that rather than digging out the values each time. It makes the code a lot easier to read.
```suggestion
static void
save_invalidation(const CaggInvalidationState *state, const Invalidation *invalidation)
{
	int32 cagg_hyper_id = state->cagg.data.mat_hypertable_id;
	TupleDesc tupdesc = RelationGetDescr(state->cagg_log_rel);
	HeapTuple refresh_tup;

	refresh_tup = create_invalidation_tup(tupdesc,
										  cagg_hyper_id,
										  invalidation->modification_time,
										  invalidation->lowest_modified_value,
										  invalidation->greatest_modified_value);
	tuplestore_puttuple(state->invalidations, refresh_tup);
	heap_freetuple(refresh_tup);
}
```